### ISSUE TITLE
508-defect-4: Confirmation modal would benefit from a unique label and may have several bugs

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#54721]
### Point of contact

joshkimux

### Severity level

3, Moderate. Should be fixed in 1-3 sprints post-launch.

### Details

The confirmation modal currently has the following potential bugs:

<img width="889" alt="Screenshot 2023-03-07 at 12 58 23 PM" src="https://user-images.githubusercontent.com/14154792/223508717-602635c7-2ab3-4c1b-a71a-3e5836f9d21a.png">

> If the user clicks back, but nothing is entered on the page, the confirmation modal is still triggered

<img width="626" alt="Screenshot 2023-03-07 at 1 02 37 PM" src="https://user-images.githubusercontent.com/14154792/223509654-6b731a1c-8a5a-48b5-8111-480e51e6d7d7.png">

> The confirmation modal does not use the unique name of the facility the user enters

The following accessibility issue is a `defect-level-4` (and possible 3) given there are multiple facilities a user can enter. For Veterans with cognitive disabilities or with short term memory (or those who have been distracted), it may not be clear what facility they are confirming.

### Reproduction steps

1. [Go to the VA Medical Records page and add a location](https://staging.va.gov/decision-reviews/supplemental-claim/file-supplemental-claim-form-20-0995/supporting-evidence/va-medical-records?index=1)
2. Click "back" and confirm the modal is triggered despite nothing being entered on the page to save in the first place

### Proposed solution or next steps

| Data     | Forward     | Back               | Add another      |
|----------|-------------|--------------------|------------------|
| Complete | Next page   | Prev page          | New page (empty} |
| Empty    | Focus error | Prev page & remove | Focus error      |
| Partial  | Focus error | Modal & Prev page  | Focus error      |

### References, articles, or WCAG support

1.
2.
3.
...

### Type of issue

- [X] Screenreader
- [ ] Keyboard
- [X] Focus
- [ ] Headings
- [ ] Color
- [ ] Zoom
- [ ] Semantics
- [ ] Axe-core
- [ ] Something else

**Comment by eileen-coforma** [comment]
The intent for the modal is to give the user who has a partially saved form an option to go back and reference what they've added before, in case they need that to continue filling out the information.

Not sure about the third proposed solution, if there is no error on the existing page, and they've fully filled out the form with the intent the add the location, the modal wouldn't need to show up.

For proposed next solutions:
-  If errors persist on the page, **continue** should send focus to the first error and not trigger the modal **Yes**
-  If errors persist on page, **back** would trigger modal 
-  If the user clicks **back**, but no content is added to the page, the confirmation modal should not be triggered **Yes**
-  If the user clicks **continue**, and no error exist on the page, the confirmation modal should trigger using the unique name of the facility e.g. "Do you want to keep [name of facility]" **Not necessary, it would continue as normal**

@joshkimux You're thoughts?

CC: @Mottie

**Comment by joshkimux** [comment]
You word it beautifully @eileen-coforma , agree on all points-- and my third proposed solution does not make sense in hindsight 💯 

Will update the ticket to match

**Comment by Mottie** [comment]
Additional scenarios... when the user adds another location, then:
- If the user clicks back, but no content is added to the page, should we keep the empty entry?
- If the user clicks continue, but no content is added to the page, should the modal show then? Or only focus on the first error?

**Comment by joshkimux** [comment]
> If the user clicks back, but no content is added to the page, should we keep the empty entry?
Good question... If it's empty, does it exist? / should it exist? That wouldn't be an a11y issue though, but worth roping in @eileen-coforma to noodle on.

> If the user clicks continue, but no content is added to the page, should the modal show then? Or only focus on the first error?
My gut says focus on the first error, as we wouldn't want them proceeding with an empty state. @eileen-coforma to keep me honest here.

**Comment by Mottie** [comment]
Is it ok if I update the modal "Yes" button to say "Yes, keep this location"... I always feel like I have to think about which button to click when it pops up

**Comment by Mottie** [comment]
Ok, one more scenario... if the user clicks "Add another location", what do we do?

* current entry empty - focus on error?
* entry with error(s) - open modal, or focus on error?

**Comment by eileen-coforma** [comment]
> Is it ok if I update the modal "Yes" button to say "Yes, keep this location"... I always feel like I have to think about which button to click when it pops up

Let's avoid changing button language until we do more testing. The button might be harder to scan/read if it's longer. We added to the No only because it does additional action of removing the location.

> Ok, one more scenario... if the user clicks "Add another location", what do we do?

Focus on the error. A person shouldn't click 'add another location' if they didn't finish entering the info for the current location, it's akin to pressing 'continue' when they didn't fill in required fields.

**Comment by Mottie** [comment]
Ok to summarize:

| Data     | Forward     | Back               | Add another      |
|----------|-------------|--------------------|------------------|
| Complete | Next page   | Prev page          | New page (empty} |
| Empty    | Focus error | Prev page & remove | Focus error      |
| Partial  | Focus error | Modal & Prev page  | Focus error      |

So the modal _only_ shows when we're going back to the previous page with partial data?

Does this all sound right?

**Comment by Mottie** [comment]
Ready for testing in staging

**Comment by eileen-coforma** [comment]
Found a bug: https://www.loom.com/share/e796876da8cf4bf3af4e199ecf4fdd4b

Under private medical records.
- Add a new location
- Enter value in the date (incomplete) and pop inline error
- Hit back
- No modal
- Add a new location, and the information is gone

**Comment by Mottie** [comment]
Ok, I found the source of that bug... working on it now!

**Comment by Mottie** [comment]
Please check that the bug has been fixed.

**Comment by joshkimux** [comment]
<img width="804" alt="Screenshot 2023-03-29 at 4 58 15 PM" src="https://user-images.githubusercontent.com/14154792/228665871-e794207a-2c01-4671-be4c-fb96509373d3.png">

Modal appears after recreating @eileen-coforma 's bug

**Comment by joshkimux** [comment]
https://user-images.githubusercontent.com/14154792/228667016-49f35709-1d6c-454e-bc15-23e35e112066.mov

One teeny follow up: @eileen-coforma and @Mottie when errors exist and then content is deleted from them, should we still show the modal?

**Comment by eileen-coforma** [comment]
I agree, that should follow the empty state scenario



Data | Forward | Back | Add another
-- | -- | -- | --
Empty | Focus error | Prev page & remove | Focus error

**Comment by joshkimux** [comment]
I'll try again! 🙇

**Comment by joshkimux** [comment]
@Mottie tested again and can confirm that:
- if a user tabs all the way through without inputting anything and clicks "back", it goes to the previous page and removes
- if a user inputs something into the memorable date fields, tabs out, tabs back in and deletes it, then tabs to "back"-- the confirmation modal will appear


https://user-images.githubusercontent.com/14154792/229613436-95e043cb-906a-4d92-af10-8dee0bd07d23.mov

**Comment by Mottie** [comment]
I see the "No issues were selected" message... this means that nothing was selected on the `/contestable-issues` page. Maybe that's why?

**Comment by Mottie** [comment]
Wait, maybe there is an issue... let me investigate further

**Comment by Mottie** [comment]
Fix merged in, will be available for [testing in staging soon](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/frontend-support-dashboard/)

**Comment by saderagsdale** [comment]
@Mottie Is this ready to test?

**Comment by joshkimux** [comment]
@saderagsdale testing now!

**Comment by joshkimux** [comment]
This is fixed! @Mottie is the 🐐 as always ❤️ 

https://user-images.githubusercontent.com/14154792/230086146-e37df901-78a9-406c-bd54-5b00c4749bb9.mov
